pkg/reconcilers/targetref_reconciler.go

```diff
@@ -196,6 +197,44 @@ func (r *TargetRefReconciler) DeleteTargetBackReference(ctx context.Context, pol
 	return nil
 }
 
+// GetAllGatewayPolicyRefs returns the policy refs of a given policy kind from all gateways managed by kuadrant.
+// The gateway objects are handled in order of creation to mitigate the risk of non-idenpotent reconciliations based on
+// this list of policy refs; nevertheless, the actual order of returned policy refs depends on the order the policy refs
+// appear in the annotations of the gateways.
+// Only gateways with status programmed are considered.
+func (r *TargetRefReconciler) GetAllGatewayPolicyRefs(ctx context.Context, policyRefsConfig common.PolicyRefsConfig) ([]client.ObjectKey, error) {
+	var uniquePolicyRefs map[string]struct{}
+	var policyRefs []client.ObjectKey
+
+	gwList := &gatewayapiv1beta1.GatewayList{}
```

Comment:
This is the only thing I would like to rise.

The previous implementation was designed with a premise: try to avoid "Get all gateways" or "Get all routes" from the cluster. 

Something that deserves further consideration, maybe for another PR